Fixed #23658 -- Provides the password to PostgreSQL from "dbshell" command

nirgal opened the pull request:
The password from settings.py, if present, is written in a temporary
.pgpass file whose name is given to psql using PGPASSFILE environment
variable.

auvipy commented:
are you missing any test?

nirgal commented:
Hi

> Are you missing any test?

I'm not sure I understand the question.

That patch is in a Works-For-Me state:
It has been tested with python2 and python3, on Debian GNU/Linux.
It requires postgresql > 9 to work. I think PGPASSFILE env is ignored before that version. But then there is nothing we can do, and dbshell command will fall back to ask the user password, as it did before, which is OK.
Care is being taken not to generate that file is there is no password, since there may be a preexisting pgpass with a solution.
I tested unicode data in the settings, and it works ok with python3. There probably is an issue with python2 there. I double check now (stay tuned).

But maybe you are implying I should have a unit test of some kind?

All automatic tests are now runing ok. It took me a while to realize isort do want an extra \n after the from __future__ import, but this has been fixed yesterday...

mjtamlyn commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -7,13 +11,40 @@ class DatabaseClient(BaseDatabaseClient):
+            # Create temporary .pgpass file
+            # It's deleted automagically on sys.exit
+            temp_pgpass = tempfile.NamedTemporaryFile(mode='w+')
+            print(
```
Better to use the file handler's write than `print`.

nirgal replied:
Actually, I tried write() first, but then I hit a bunch of problems regarding byte/string encoding.
I always use utf-8 on Debian GNU/Linux, but I'm not sure how to proceed so that it works everywhere.
We could use force_byte or whatever, but if the locale is not unicode for exemple iso-8859-15, I'm affraid the utf-8 encoded password will not be accepted by psql that will be runing with default environement.
I tested print() for both python2 & 3, with a 'é' in the password, and it seems ok in all cases.
I'm not sure about windows, I don't have access to one.
After considering all the tests that might be required to have it right, including testing process locale and differences between python 2 & 3, I decided to give up that approach.
print() just works, as far as I can tell. I believe it is less ugly that the alternative...

I could use some help if you think using print is unacceptable, please.

mjtamlyn commented:
We do currently have no tests I can find for `dbshell`. It would be nice to have some... Also is there any way we can safely remove the temporary file afterwards?

mjtamlyn commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -7,13 +11,40 @@ class DatabaseClient(BaseDatabaseClient):
     executable_name = 'psql'
 
     def runshell(self):
+        def escape_pgpass(txt):
```
This function shouldn't be defined inside the method.

nirgal replied:
> [escape_pgpass] shouldn't be defined inside the method

Agreed.
I'm not sure about the best convention to use:
Should it be a staticmethod of the class?
Or a _escape_pgpass on the gobal level?
Or just a regular escape_pgpass on the global level?

aaugustin replied:
Any of these three solutions works for me :-)

nirgal replied:
Fixed in commit c21b2e19146985e949e5e7353e0f8cf4b11a7d0b

nirgal commented:
The temporary file is removed, as stated in the comments.
It tested it on Debian GNU/Linux.
"pydoc tempfile.NamedTemporaryFile" says the removal is automatic.

aaugustin commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -7,13 +11,40 @@ class DatabaseClient(BaseDatabaseClient):
+        if passwd:
+            # Create temporary .pgpass file
+            # It's deleted automagically on sys.exit
+            temp_pgpass = tempfile.NamedTemporaryFile(mode='w+')
```
Most likely this will not work on Windows because files created with `NamedTemporaryFile` cannot be reopened on Windows (which defeats the whole purpose of naming them in the first place -- I have no idea why `NamedTemporaryFile` even exists on Windows).

I'm not saying this is blocking the merge because I don't think we have that many users of PostgreSQL on Windows, but I thought I'd bring it up in case someone wants to check.

nirgal replied:
You are correct. I added comments in commit 6b6aef85e42e5d7f68e3f4f8b238d5de985a7a0d, including a link to python bug tracker http://bugs.python.org/issue14243

> I'm not saying this is blocking the merge because I don't think we have that many users of PostgreSQL on Windows, but I thought I'd bring it up in case someone wants to check.

I'm not sure I want to write a complex work around for a python related bug that will eventually by fixed upstream.

Please someone ping me if you think this is required.

nirgal commented:
Hi

The "flake8" test is showing an error:
http://djangoci.com/job/flake8/1271/console
> stderr: fatal: unable to connect to github.com:
> github.com[0: 192.30.252.128]: errno=Connection timed out
I think it was because the test attempted to run during a ddos ( https://twitter.com/githubstatus/status/581298353771802624 )

How can I relaunch it?

MarkusH commented:
check flake8

MarkusH commented:
There you go :smile:

nirgal commented:
Last commit should take care of the issue with microsoft temporary files. :)

MarkusH commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,63 @@
+        passwd = settings_dict.get('PASSWORD', '')
+
+        if user:
+            args += "-U", user
```
can you explicitly wrap them in brackets: `args += ["-U", user]` please. That makes it clearer to understand the code.

nirgal replied:
done

MarkusH commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,63 @@
+                passwd,
+                file=temp_pgpass,
+                sep=':',
+                )
```
Unindent by one level, please.

nirgal replied:
done

MarkusH commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,63 @@
+        if passwd:
+            # Create temporary .pgpass file. This requires PostgreSQL >= 8.1.
+
+            temp_pgpass = tempfile.NamedTemporaryFile(mode='w+', delete=False)
```
I think `NamedTemporaryFile()` supports context managers, doesn't it. You should be able to use

```python
with tempfile.NamedTemporaryFile(...) as temp_pgpass:
    # Do some magic
```

nirgal replied:
Hi Markus.

Because of the issue of NamedTemporaryFile that can't be reopen on windows, we need to unlink the file manually. See the comment in the code.

Moreover, the file is created only if there is a password. It might not be the case. This might happen because the system is set up in a way to not require it when using a socket for example (host is empty then); Or because the password is available in an existing .pgpass file, so that we don't want to create another one, and let postgres use the default.

For these 2 reasons, it's quite difficult to use `with`: subprocess must be called whether there is a password or not. unlink must also be called in the two cases, because of the windows buggy tempfile.

So I don't really understand how to use with in these cases. That would be a lot of duplicate code. Do you have a idea?

MarkusH replied:
Ah, got it. In that case you should rely on try-finally.

```python
temp_pgpass = None
try:
    if passwd:
        temp_pgpass = tempfile.NamedTemporaryFile(...)
        ...
    subprocess.call(args)
finally:
    if temp_pgpass and os.path.exists(temp_pgpass.name):
        os.unlink(temp_pgpass.name)
```

nirgal replied:
Yes, good catch. `call` might raise an exception if psql is not in the PATH, and we wouldn't want tmp to be filled with obsolete files.
Commited in my branch.

MarkusH commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,65 @@
 
 
+def _escape_pgpass(txt):
+    'Escape a non-terminal fragment of a postgresql .pgpass file'
```
User triple double-quotes (`"""`) for docstrings, please.

MarkusH commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,65 @@
+        try:
+            if passwd:
+                # Create temporary .pgpass file. This requires PostgreSQL >= 8.1.
+
```
Drop this empty line

MarkusH commented:
Can you add a release note to `docs/releases/1.9.txt` at https://github.com/django/django/blob/master/docs/releases/1.9.txt#L189 please.

MarkusH commented:
Can you squash all commits when you're done with the remaining annotations, please.

MarkusH commented:
buildbot, ok to test.

nirgal commented:
Hi Markus

- User triple double-quotes (""") for docstrings done. I added a dot at the end.
- Empty line after "Create temporary .pgpass file." comment removed.
- Release note added. (English is not my native language, fixes welcomed).
- All commits squashed in one.

If you see anything else trivial you'd like to change, fell free to do it, with or without my name.

nirgal commented:
check default

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -201,6 +201,10 @@ Management Commands
 * The :djadmin:`createcachetable` command now has a ``--dry-run`` flag to
   print out the SQL rather than execute it.
 
+* When using the PostgreSQL backend, the :djadmin:`dbshell` command no longer
```
... command can connect to the database using the password from your settings file (instead of requiring it to be manually entered).

timgraham commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,66 @@
+        temp_pgpass = None
+        try:
+            if passwd:
+                # Create temporary .pgpass file. This requires PostgreSQL >= 8.1.
```
As Django requires PG 9+, I don't think the comment about PG >= 8.1 is needed.

timgraham commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,66 @@
+                # Create temporary .pgpass file. This requires PostgreSQL >= 8.1.
+                temp_pgpass = tempfile.NamedTemporaryFile(mode='w+', delete=False)
+
+                # Note that print supports non ascii7 characters in strings on
```
Not sure the relevance of this comment?

timgraham commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,66 @@
+        passwd = settings_dict.get('PASSWORD', '')
+
+        if user:
+            args += ["-U", user]
```
use single quotes for consistency?

timgraham commented on django/db/backends/postgresql_psycopg2/client.py:
```diff
@@ -1,19 +1,66 @@
+                    sep=':',
+                )
+
+                # We need to close the file and remove it manually because
```
Did you have a look if `django.core.files.temp.NamedTemporaryFile` might help here?

timgraham commented:
Maybe some tests could be added with the help of mock?

nirgal commented:
Tim:
- Updated release note, thanks.
- Drop comment about postgres > 8.1, this was stupid now that it's so far back.
- Made comment about print smaller. I did missed the issues about the non-ascii7 password with the different combinations of locales and python versions at first, so I do believe this is nice to have. It's now half a line, I hope you like it.
- Use single quotes in arguments for consistency.
- Switch to django.core.files.temp.NamedTemporaryFile. I did not know that one and that was exactly what was needed, thanks :) I remove the delete=False in its call. Then I added flush=True to print, droped the close just after, and replace the final to unlink by a simple close. Tested on Linux.
- Regarding the possibility to add more tests with the help mock, I currently don't have the knowledge to do that (Never used it, shame on me). I think the main point however is to check the temporary file with the password doesn't stay around. But after a single manual test, I think that burden should now be on django.core.files.temp.NamedTemporaryFile.

I started editing the commit message to add thanks, but there are so many people helping that they would be too long or frustrating for people not quoted. Thank you all.

nirgal commented:
I just refreshed that branch against master. It's ready to merge if you have no further comment.

timgraham commented:
We are waiting for someone to investigate the possibility of adding tests.

nirgal commented:
I tried to add a test, similar to what is done with mysql. (see last commit)
Is that what you had in mind?

timgraham commented on tests/dbshell/tests.py:
```diff
@@ -70,4 +75,27 @@ def test_ssl_certificate_is_added(self):
+        return MySqlDatabaseClient.settings_to_cmd_args(connection_settings)
+
+
+class PostgreSqlDbshellCommandTestCase(SimpleTestCase):
```
Could you put the new tests in a new file: `test_postgresql_psycopg2.py`? I'll rename the existing file to `test_mysql.py` in a separate commit.

timgraham commented on tests/dbshell/tests.py:
```diff
@@ -70,4 +75,27 @@ def test_ssl_certificate_is_added(self):
+                self.last_subprocess_args,
+                ['psql', '-U', 'someuser', '-h', 'somehost', '-p', '444',
+                 'somedbname'])
+            self.assertIn('PGPASSFILE', self.last_subprocess_env)
```
could we read the file and verify its contents?

nirgal commented:
Hi.

I moved the tests in a separate file as requested.

I also read the content of the file, now. And this is quite complex because we want this to work on both python2 and python3, and both with unicode and other locales.
I took some arbitrary username/password (rôle/sésame) and I skip one test is the current locale doesn't support these characters (I tried with Japanese SJIS: test is skiped).
I ran the tests with the four combinations LANG=fr_FR vs LANG=fr_FR.UTF-8, and python2.7 vs python3.4, it all looks fine here.

I also added more tests and could fix additional problems:
- The environment must be cleaned after calling dbshell, hence the del os.environ['PGPASSFILE']
- I tested a password with a column, and it needs to be escaped too. Code was fixed. :)

I have one question: Is "import locale" supported on all platform? No problem here on linux, but I'm not sure globally.

nirgal commented:
check default

timgraham commented:
We import `locale` elsewhere at the top level of Django, so should be fine.  Here are some edits: http://dpaste.com/0NTP1PK

timgraham commented on tests/dbshell/test_postgresql_psycopg2.py:
```diff
@@ -0,0 +1,106 @@
+                b'somehost:444:somedbname:some\\\\user:some\\\\password\n'))
+
+    def test_accent(self):
+        encoding = locale.getpreferredencoding()
```
under what circumstances do we need this? My system has 'UTF-8', so it's not very exciting as that's the default for these functions.

nirgal replied:
I tried installing with a virtual Debian machine, I did a "dpkg-reconfigure locales" and I selected locale fr_FR whose default encoding is ISO-8859-1. This is quite unusual since most people use fr_FR.UTF-8 nowadays. I did that before installing postgres.
The pgpass temporary file needs to be encoded using system locale.
"rôle" must be encoded using 5 bytes on utf-8 systems, it must be encoded using 4 bytes on iso-8859-1 systems.

timgraham replied:
Thanks for the explanation. I added a brief comment about that.

timgraham commented on tests/dbshell/test_postgresql_psycopg2.py:
```diff
@@ -0,0 +1,106 @@
+                u'somehost:444:somedbname:%s:%s\n' % (username, password),
+                encoding=encoding)
+        except UnicodeEncodeError:
+            self.skipTest("Your locale can't run this test with python2.")
```
I removed this in my edits since the test worked fine for me, but I depending on the answer to the question above, I guess it might need to be added back.

nirgal replied:
As I explained earlier, I tried forcing encoding=sjis. This is a Japanese encoding.
There is no way to write "rôle" in a file using that encoding. "ô" is just not part of the alphabet.
I skip the test when that happens.

timgraham commented:
merged in b64c0d4d613b5cabedbc9b847682fe14877537de, thanks!
